packages/yoroi-extension/app/api/ada/transactions/shelley/transactions.js

```diff
@@ -433,6 +433,70 @@ export function newAdaUnsignedTxFromUtxo(
   |}>,
   allowNoOutputs: boolean,
   metadata: RustModule.WalletV4.TransactionMetadata | void,
+): V4UnsignedTxUtxoResponse {
+  /*
+    This is an ad-hoc optimization for one specific senario:
+    If the input is enough to cover the output and the fee, but the remaining amount
+    is less than the minimum UTXO amount (1 ADA as of now), then
+    `txBuilder.add_change_if_need` will "burn" the remaining as fee instead of adding
+    a change.
+    The optimization is to include one extra UTXO input to force the change amount
+    to be larger than the minimum UTXO amount.
+  */
+  const result = _newAdaUnsignedTxFromUtxo(
+    outputs,
+    changeAdaAddr,
+    utxos,
+    absSlotNumber,
+    protocolParams,
+    certificates,
+    withdrawals,
+    allowNoOutputs,
+    metadata,
+    false,
+  );
+  const fee = result.txBuilder.get_fee_if_set();
+
+  const resultWithOneExtraInput = _newAdaUnsignedTxFromUtxo(
+    outputs,
+    changeAdaAddr,
+    utxos,
+    absSlotNumber,
+    protocolParams,
+    certificates,
+    withdrawals,
+    allowNoOutputs,
+    metadata,
+    true,
+  );
+  const feeWithOneExtraInput = resultWithOneExtraInput.txBuilder.get_fee_if_set();
+
+  if (feeWithOneExtraInput && fee && feeWithOneExtraInput.compare(fee) < 0) {
+    return resultWithOneExtraInput;
+  }
```

Comment:
Nice! Good check! I was thinking we gonna need to do some minimum threshold check and only add the additional input if the extra burn is above that threshold, but this approach is super straightforward and way better!